### ISSUE TITLE
order chaninng

### ISSUE THREAD

**Issue by Fux002** [quickfix/quickfix#286]
Hi can someone help me to change the order of the different tag in a fix message generated with quickfix?

thx a lot!

**Comment by Ambal** [comment]
hi, @Fux002 . You can only change the order of the tags in the repeating group. Unfortunately, you need to regenerate/manually change code of pre-generated FIX message classes in order to achieve your goal.

**Comment by Fux002** [comment]
Ok great thank you for the reply!

Can you perhaps tell me where I can find this in the scripts.

Just to be sure this modification is not possible in the data dictionary?

Kind regards,

Fons

**Comment by Ambal** [comment]
Hi, @Fux002 . There is no way to reorder field tags from client code. The code generators are in "quickfix/spec" folder, they will generate message classes based on data dictionaries you'll provide. Or you can alter the generated classes in "quickfix/include" folder - you will have to modify the FIX::message_order(XXXXX) content.

`FIX::Group(1018,1019,FIX::message_order(1019,1050,1051,1052,0)) {}`

**Comment by Fux002** [comment]
Hi Victor thank you for this and apologies I’m new to the protocol.

I have a data dictionary from the broker I want to send messages to. Why are my messages not in the correct order. I am reluctant to change the source code as I deal with multiple brokers and this will not be consistent for all of them.

What’s wrong with the dictionary?

**Comment by Ambal** [comment]
Hi, @Fux002 . Now I don't quite understand what your problem is. Do you have issues with field ordering for particular messages? Or do you send messages in the incorrect order to broker?

**Comment by Fux002** [comment]
Thank you for the kind reply,

If this is the data dictionary i work with:
[image: image.png]

The NoRelatedSym in the broker docs is not a group but it id in the DD. Di
i make that a group?

How do I set the groups correctly for now I have:

        message = fix.Message()
        header = message.getHeader()
        header.setField(fix.MsgType("V")) #35
        header.setField(57, 'MD')

        group = fix.MarketDataRequest().NoMDEntryTypes()
        group.setField(269, "0")
        message.addGruop(group)

        message.setField(262, "99999999")
        message.setField(146, "1")
        message.setField(55, "GBP/USD")
        message.setField(263, "1")
        message.setField(264, "1")
        message.setField(265, "0")
        message.setField(267, "2")
        message.setField(1, "10476936")
        fix.Session.sendToTarget(message, self.sessionID)

thx a lot!

**Comment by Fux002** [comment]
ok got it! starting to get the hang of it slowly :)

message = fix.Message()
        header = message.getHeader()
        header.setField(fix.MsgType("V"))

        message.setField(262, "99999")
        message.setField(263, "1")
        message.setField(264, "1")
        message.setField(265, "1")
        message.setField(267, "1")

        group = fix44.MarketDataRequest().NoMDEntryTypes()
        group.setField(269, "0")
        message.addGroup(group)

        symbol = fix44.MarketDataRequest().NoRelatedSym()
        symbol.setField(55, "GBP/USD")
        message.addGroup(symbol)

        message.setField(1, "10476936")
        fix.Session.sendToTarget(message, self.sessionID)

**Comment by leleftheriades** [comment]
Since you are new, I am suggesting a more readable use of the API
```
FIX44::MarketDataRequest request(
    FIX::MDReqID("reqId"), 
    FIX::SubscriptionRequestType(FIX::SubscriptionRequestType_SNAPSHOT_PLUS_UPDATES),
    FIX::MarketDepth(16)
);

FIX44::MarketDataRequest::NoMDEntryTypes bid;
bid.set(FIX::MDEntryType(FIX::MDEntryType_BID));
request.addGroup(bid);

FIX44::MarketDataRequest::NoMDEntryTypes offer;
offer.set(FIX::MDEntryType(FIX::MDEntryType_OFFER));
request.addGroup(offer);

request.setField(FIX::MDUpdateType(FIX::MDUpdateType_FULL_REFRESH));

FIX44::MarketDataRequest::NoRelatedSym symbol;
symbol.set(FIX::SecurityID("EUR/USD"));
symbol.set(FIX::SecurityIDSource(FIX::SecurityIDSource_EXCHANGE_SYMBOL));
request.addGroup(symbol);
```

**Comment by Fux002** [comment]
Hey thank you!

I got it all figured out. Working with groups wasn t clear to me but that’s all fine. Thank you!

One more question. Does the library reconnect itself when needed or do I implement this logic form for example the heartbeat.

Thank you,

Fons

**Comment by leleftheriades** [comment]
the library will automatically reconnect and manage heartbeats

**Comment by Fux002** [comment]
Great thx!

**Comment by Fux002** [comment]
Good morning,

One of my counter parties on fix require me to have a tag 1128 on the login message. I don’t seem to be able to set this tag. Can you please help me with this?

Kind regards,

Fons

**Comment by leleftheriades** [comment]
```
static bool isLogonMessage(const FIX::Message & message) 
{
    return message.getHeader().getField(FIX::FIELD::MsgType) == FIX::MsgType_Logon;
}

void toAdmin(FIX::Message & message, const FIX::SessionID & sessionId)
{
    if (isLogonMessage(message)) {
       message.setField(FIX::ApplVerID("1.00"));
    }
}
```

**Comment by Fux002** [comment]
Thank you for the reply.

I’m not sure it works to set the variable like that.

Please look at this problem it’s similar.

http://quickfix-j.364392.n2.nabble.com/1128-Field-Missing-td7116521.html

I’m reluctant to start changing the source files. What is the best option for me please.

Have a great weekend,

Fons

**Comment by lefterise** [comment]
I am talking about your source files not quickfix's. 
You can make your source read from the settingsfile if you like.
Quickfix [does not](https://github.com/quickfix/quickfix/blob/master/src/C%2B%2B/Session.cpp#L673) populate that tag for you.

**Comment by Fux002** [comment]
Apologies for my questions if they are slightly basic.

I tried to add it to the “toAdmin” block this doesn’t seem to work.

What would be my source file please? How do I make it read from the settings file. I was of the impression that adding the tag to the data dictionary would mean its added to the login message by default.

Thank you for the support so far.

Fons

**Comment by lefterise** [comment]
The data-dictionary is only used for validation. It doesn't add fields.

To use quickfix you need to implement the Application interface. 
This interface allows you to intercept:
- outgoing admin messages (toAdmin)
- outgoing messages (toApp)
- incoming admin messages (fromAdmin)
- incoming messages (fromApp)

The logon message is sent by the library, with specific fields populated. Your goal is to intercept that message and attach a field to it. You can do this by checking if the message in toAdmin is a logon message and attaching the field. 

Note that there are 3 possible places to attach a field: The header, the body and the trailer.
Maybe ApplVerID should be attached in the header, so try the following:

message.getHeader().setField(FIX::ApplVerID("1.00"));
